libraries/Microsoft.Bot.Builder.Azure/AzureTableStorage.cs

```diff
@@ -3,164 +3,369 @@
+                var type = Entity.GetType();
+                var typeQualifiedName = type.AssemblyQualifiedName;
+                properties.Add("__type", EntityProperty.GeneratePropertyForString(typeQualifiedName));
+
+                return new DynamicTableEntity(Key.PartitionKey, Key.RowKey)
                 {
-                    var result = await this.Table.ExecuteAsync(TableOperation.Replace(entity)).ConfigureAwait(false);
+                    ETag = ETag,
+                    Properties = properties
+                };
+            }
+
+            public static StoreItemEntity AsStoreItemEntity(TableResult tableEntity)
+            {
+                // Create instance of proper type
+                var dynamicTableEntity = (DynamicTableEntity)tableEntity.Result;
+                var type = Type.GetType(dynamicTableEntity.Properties["__type"].StringValue);
+                var properties = dynamicTableEntity.Properties;
+                var value = CreateInstaceOf(type);
+
+                // Set object properties
+                if (IsAnonymousType(type))
+                {
+                    // Anonymous Type not supported - Anonymous Types may not have setters as values are passed in the constructor
+                    // Convert to dynamic/ExpandoObject instead
+                    var expando = new ExpandoObject() as IDictionary<string, Object>;
+
+
+                    // TODO: Handle nested objects (e.g.: Parent_Child_Prop1)
+                    foreach (var prop in properties)
+                    {
+                        var propKey = prop.Key;
+                        var propValue = prop.Value.PropertyAsObject;
+                        expando[propKey] = propValue;
+                    }
+
+                    value = expando;
                 }
                 else
                 {
-                    throw new Exception("etag empty");
+                    // POCO entity
+                    TableEntity.ReadUserObject(value, properties, new OperationContext());
                 }
+
+                // IStoreItem? apply Etag
+                if (value is IStoreItem iStoreItem)
+                {
+                    iStoreItem.eTag = tableEntity.Etag;
+                }
+
+                return new StoreItemEntity(new EntityKey(dynamicTableEntity.PartitionKey, dynamicTableEntity.RowKey), value);
             }
-        }
 
-        protected class StoreItemEntity : TableEntity
-        {
-            private static JsonSerializerSettings serializationSettings = new JsonSerializerSettings()
+            private static void Flatten(IDictionary<string, EntityProperty> properties, object entity, string prefix = "")
             {
-                // we use all so that we get typed roundtrip out of storage, but we don't use validation because we don't know what types are valid
-                TypeNameHandling = TypeNameHandling.All
-            };
+                if (IsAnonymousType(entity.GetType()))
```

Comment:
Interesting. Did you do this because you found that the `EntityPropertyConverter` didn't support them out of the box? I honestly don't know if we want to be in the business of trying to do our own serialization just to support anonymous types. I honestly don't think we _need to_ support them. I think if there's a sample out there using them already it's contrived. Anonymous types are highly localized to one method, at bare minimum you need to use a tuple to pass between methods. 

No other persistence frameworks I'm aware of support anonymous types. Serialization frameworks, yes, but not persistence. Also, if this provider supports it, then suddenly all providers would need to do the same or the abstraction doesn't work. 

My vote would be to remove support for anonymous types. If not for the reason I've already laid out, then for the fact that most of the other comments you'll see on this PR are about the implementation of support for them. @cleemullins What do you think?

Reply:
While I *LOVE* that you added support for anonymous types, I'm fine removing that support if it means we can use existing, debugged, standard code from libraries we're already using. 

My inclination would be to punt on this and use the existing Azure bits.

Reply:
Makes sense, either support it across the SDK or not =)
Dropping Anonymous Types support and using TableEntity.Flatten only